### PR TITLE
Walker407/hotdog roller

### PR DESCRIPTION
**Pull request by walker407**

### CODE REVIEW

File: code/Cinema.csproj [hunk 1]
```diff
@@ -0,0 +1,55 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (Remscar) [inline code/Cinema.csproj:1]: Remove the csproj file, we don't check that in anymore

File: code/entities/cooking/hotdog-roller/components/interactions/switches/BaseSwitch.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
+﻿using Sandbox;
+
+namespace Cinema;
+
+public partial class BaseSwitch : EntityComponent<HotdogRoller>
```
Comment (Remscar) [inline code/entities/cooking/hotdog-roller/components/interactions/switches/BaseSwitch.cs:5]: You could refactor `BaseSwitch` `LeftSwitch` and `RightSwitch` into a single class, `HotdogRollerSwitch`

The only difference between these classes is `LeftSwitch` sets the parameter (toggle_left) and `RightSwitch` sets the parameter (toggle_right)

Comment (walker407) [reply code/entities/cooking/hotdog-roller/components/interactions/switches/BaseSwitch.cs:5]: That's definitely something I want to rethink / refactor. I originally had it as one class, but the way the model is setup currently, I found it to be cleaner to have two separate classes for now

File: code/entities/cooking/hotdog-roller/extensions/HotdogRoller.State.cs [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    {
+        TransitionStateTo(State.BothOff);
+    }
+
+    /// <summary>
+    /// Updates the power state when one of the power switches changes state
+    /// </summary>
+    private void UpdatePowerState()
+    {
+        if (!LeftSwitch.IsOn() && !RightSwitch.IsOn())
```
Comment (Remscar) [inline code/entities/cooking/hotdog-roller/extensions/HotdogRoller.State.cs:27]: There is a way to simplify this function and have only 3 if statements (1 nested in another)

Using else statements are appropriate in this case too.

File: code/entities/cooking/hotdog-roller/extensions/HotdogRoller.State.cs [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+                break;
+            case State.FrontOn:
+                HandleFrontOnState();
+                break;
+        }
+    }
+    /// <summary>
+    /// Handles transition to new state
+    /// </summary>
+    private void TransitionStateTo(State state)
```
Comment (Remscar) [inline code/entities/cooking/hotdog-roller/extensions/HotdogRoller.State.cs:80]: As mentioned in another comment, you should refactor all this logic out of here and into the `Roller` component.

File: code/entities/cooking/hotdog-roller/components/interactions/roller/Roller.cs [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    /// Attaches hotdog by attachment point
+    /// </summary>
+    private void AttachEntity(string attach, Entity ent)
+    {
+        if (Entity.GetAttachment(attach) is Transform t)
+        {
+            ent.Transform = t;
+
+            ent.Position += (Vector3.Up * 0.5f);
+            ent.Position += (Vector3.Left * 0.5f);
```
Comment (Remscar) [inline code/entities/cooking/hotdog-roller/components/interactions/roller/Roller.cs:153]: Just using `.Up` and `.Left` won't work with rotated entities. Use Orientation vectors

File: code/entities/cooking/hotdog-roller/components/interactions/switches/BaseSwitch.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+        TransitionStateTo(state);
+    }
+
+    /// <summary>
+    /// Toggles state between State On and State Off
+    /// </summary>
+    public void TogglePos()
+    {
+        if(SwitchState == State.Off)
```
Comment (Remscar) [inline code/entities/cooking/hotdog-roller/components/interactions/switches/BaseSwitch.cs:22]: If you make SwitchState a boolean then you can do `TransitionStateTo(!isRolling)`

File: code/entities/cooking/hotdog-roller/extensions/HotdogRoller.State.cs [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        MachineState = state;
+
+        HandleState();
+    }
+    /// <summary>
+    /// Handles state for both rollers off
+    /// </summary>
+    private void HandleBothOffState()
+    {
+        SetMaterialGroup(0);
```
Comment (Remscar) [inline code/entities/cooking/hotdog-roller/extensions/HotdogRoller.State.cs:91]: Create an enum of the material group (if it's not already the same as the states) so you aren't using magic numbers.

File: code/entities/cooking/hotdog-roller/components/interactions/roller/Roller.cs [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    {
+        State state = (State) pos;
+
+        TransitionStateTo(state);
+    }
+
+    /// <summary>
+    /// Adds hotdog by roller id
+    /// </summary>
+    public void AddHotdog(int rollerid)
```
Comment (Remscar) [inline code/entities/cooking/hotdog-roller/components/interactions/roller/Roller.cs:33]: There is a bug with this function; If I add a hotdog while the machine is already on, it never starts rolling / steaming until i turn it off/on

**Review by Remscar — CHANGES_REQUESTED** [review]
A lot of great code in here!
The entity is really cool and it's fun to interact with, super excited for this to get in game.

A lot of my comments are around how to simplify this code to make it more efficient and easier to modify in the future.

There are also some bugs around functionality and if the machine is rotated.

I think after you address my comments it'll maybe need one more small review and then it'll be good to go!

**Review by Remscar — CHANGES_REQUESTED** [review]
A lot of great code in here!
The entity is really cool and it's fun to interact with, super excited for this to get in game.

A lot of my comments are around how to simplify this code to make it more efficient and easier to modify in the future.

There are also some bugs around functionality and if the machine is rotated.

I think after you address my comments it'll maybe need one more small review and then it'll be good to go!

**Pull request closed without merging**